W-17464439 flex security diagrams gr (#739)
mulesoft/docs-gateway: gateway/1.8/modules/ROOT/pages/flex-security-best-practices.adoc

.. <<protect-external-services>>

NOTE: Guidelines on this page are applicable to all Flex Gateway runtime environments, but diagrams are only provided for Kubernetes.

lex Gateway supports multiple runtime environments for deployment. Guidelines on this page are applicable to all environments, but details, such as diagrams, are provided 

[[secure-flex-gateway-in-anypoint-platform]]
== Secure Flex Gateway in Anypoint Platform

Always apply the principle of least privilege to Flex Gateways in Anypoint Platform. Ensure that only users that need access to Flex Gateways have it. Users that have access to Flex Gateway are able to view Flex Gateway configuration details and logs.

To learn how to manage user's access, see xref:access-management::index.adoc[].

[[secure-the-flex-gateway-registration-resource]]
== Secure the Flex Gateway Registration Resource

Flex Gateway stores registration information locally. This registration resource contains the certificate the Flex Replica uses to communicate and authenticate itself with Anypoint Platform. This diagram demonstrates how Flex Gateway communicates with Anypoint Platform via mTLS.

> Restrict access to the registration resource to ensure there are unauthorized execution of replicas from outside the organization. Specifically:

Reviewer: there are no unauthorized execution. I would not mention 'from outside the org'

[[renew-the-flex-certificate]]
== Renew the Flex Certificate 

Anypoint Flex Gateway uses a managed PKI certificate to communicate with Anypoint Platform. This certificate is generated when you register Flex Gateway in either Connected Mode or Local Mode. Renew the certificate before it's expiration date to ensure the runtime can connect to the control plane.

To learn how to renew your registration, see xref::flex-gateway-renew-certificate.adoc[].

[[update-your-flex-gateway]]
== Update your Flex Gateway

MuleSoft regularly provides security updates for Flex Gateway. Keep your Flex Gateway updated to the latest version to ensure vulnerabilities are addressed. To reduce the risk of incompatibilities, follow a safe change process and upgrade sandbox runtimes ahead of production. 

To learn more about how to update your Flex Gateway, see xref:flex-gateway-upgrade.adoc[].

[[avoid-logging-payload-data]]
== Avoid Logging Payload Data

Runtime logs are available in the runtime layer through the `stdout` of the Flex Gateway pods and also in the Anypoint control plane for Connected Mode.

> To ensure client information is not leaked if runtime logs become compromise don't log payload data in runtime logs.  

Reviewer: 'compromised'.
Also, avoid logging payload daata

= Flex Gateway Security Best Practices
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../assets/images

This page provides best practices to ensure your Flex Gateway runtime deployment meets industry standards.

> Flex Gateway deployments consist of the runtime and control plane. Information such as configuration data and logging are shared between Anypoint Platform in the control plane to Flex Gateway in the runtime plane via mTLS communication. In the runtime layer, Flex Gateway orchestrates the incoming traffic and protects the upstream service or API. Flex Gateway is registered with Anypoint Platform by using a YAML registration file. Flex Gateway uses Redis shared storage to cache data to communicate with other Flex Replicas and cache request data. This diagram demonstrates a Flex Gateway deployment:

Reviewer 2: ```suggestion
Flex Gateway deployments consist of the runtime and control plane. Information such as configuration data and logging is shared between Anypoint Platform in the control plane and Flex Gateway in the runtime plane via mTLS communication. In the runtime layer, Flex Gateway orchestrates the incoming traffic and protects the upstream service or API. Flex Gateway is registered with Anypoint Platform by using a YAML registration file. Flex Gateway uses Redis shared storage to cache data to communicate with other Flex Replicas and cache request data. This diagram demonstrates a Flex Gateway deployment:
```

image:security-best-practices.png[]

[calloutlist]
. <<secure-flex-gateway-in-anypoint-platform>>
. Protect your Flex Gateway:
.. <<secure-flex-gateway-service-pods>>
.. <<update-your-flex-gateway>>
. <<secure-the-flex-gateway-registration-resource>>
. <<renew-the-flex-certificate>>
. <<avoid-logging-payload-data>>
. <<secure-redis-shared-storage>>
. <<protect-the-flow-of-incoming-traffic>>
. <<protect-the-upstream-service>>:
.. <<protect-services-available-locally-to-flex-gateway>>
.. <<protect-external-services>>
. <<preform-regular-testing>>

NOTE: Guidelines on this page are applicable to all Flex Gateway runtime environments, but diagrams are only provided for Kubernetes.

[[secure-flex-gateway-in-anypoint-platform]]
== Secure Flex Gateway in Anypoint Platform

> Always apply the principle of least privilege to Flex Gateways in Anypoint Platform. Users that have access to Flex Gateway are able to view Flex Gateway configuration details and logs. Use users and team permissions to limit access to Flex Gateways to only those users who maintain the runtime. Provide read only access except for users who deploy or manage the gateway configuration.

Reviewer 2: ```suggestion
Always apply the principle of least privilege to Flex Gateways in Anypoint Platform. Users who have access to Flex Gateway are able to view Flex Gateway configuration details and logs. Use users and team permissions to limit access to Flex Gateways to only those users who maintain the runtime. Provide read-only access except for users who deploy or manage the gateway configuration.
```

> To learn how to manage user's access, see xref:access-management::index.adoc[].

Reviewer 2: ```suggestion
To learn how to manage access, see xref:access-management::index.adoc[].
```

[[secure-the-flex-gateway-registration-resource]]
== Secure the Flex Gateway Registration Resource

Flex Gateway stores registration information locally. This registration resource contains the certificate the Flex Replica uses to communicate and authenticate itself with Anypoint Platform. Flex Gateway communicates with Anypoint Platform via mTLS.

> Restrict access to the registration resource to ensure there are no unauthorized execution of replicas. Specifically:

Reviewer 2: ```suggestion
Restrict access to the registration resource to ensure there's no unauthorized execution of replicas. Specifically:
```

* Ensure Secrets are encrypted at rest and that least privilege access is enforced for the registration resource secret.
* Store sensitive data, such as credentials and API keys, in secure vaults, for example, AWS Secrets Manager, Azure Key Vault, or HashiCorp Vault.
* Avoid hardcoding secrets and only use environment variables for non-sensitive configurations.

> * Encrypt sensitive information in configuration files

Reviewer 2: ```suggestion
* Encrypt sensitive information in configuration files.
```

> For Kubernetes deployments, see https://kubernetes.io/docs/concepts/security/secrets-good-practices/[Good practices for Kubernetes Secrets] for Kubernetes specific best practices.

Reviewer 2: ```suggestion
For best practices for Kubernetes deployments, see https://kubernetes.io/docs/concepts/security/secrets-good-practices/[Good practices for Kubernetes Secrets].
```

[[renew-the-flex-certificate]]
== Renew the Flex Certificate 

> Anypoint Flex Gateway uses a managed PKI certificate to communicate with Anypoint Platform. This certificate is generated when you register Flex Gateway. Renew the certificate before it's expiration date to ensure the runtime can connect to the control plane.

Reviewer 2: ```suggestion
Anypoint Flex Gateway uses a managed PKI certificate to communicate with Anypoint Platform. This certificate is generated when you register Flex Gateway. Renew the certificate before its expiration date to ensure the runtime can connect to the control plane.
```

To learn how to renew your registration, see xref:flex-gateway-renew-certificate.adoc[].

[[update-your-flex-gateway]]
== Update your Flex Gateway

> MuleSoft regularly provides security updates for Flex Gateway. Keep your Flex Gateway updated to the latest version to ensure vulnerabilities are addressed. To reduce the risk of incompatibilities, follow safe change processes and upgrade sandbox runtimes ahead of production. 

Reviewer 2: ```suggestion
MuleSoft regularly provides security updates for Flex Gateway. Keep your Flex Gateway updated to the latest version to address vulnerabilities. To reduce the risk of incompatibilities, follow safe change processes and upgrade sandbox runtimes ahead of production. 
```

To learn more about how to update your Flex Gateway, see xref:flex-gateway-upgrade.adoc[].

[[avoid-logging-payload-data]]
== Avoid Logging Payload Data

> Runtime logs are available in the runtime layer through the `stdout` of the Flex Gateway pods and also in the Anypoint control plane for Connected Mode.

Reviewer 2: ```suggestion
Runtime logs are available in the runtime layer through the `stdout` of the Flex Gateway pods and in the Anypoint control plane for Connected Mode.
```

> To ensure client information is not leaked if runtime logs become compromised, avoid logging payload data in runtime logs. 

Reviewer 2: ```suggestion
To ensure compromised runtime logs don't leak client information, avoid logging payload data in runtime logs. 
```

If using other third party monitoring and analytics platforms, ensure you consider then in access reviews.

[[secure-flex-gateway-service-pods]]
== Secure Flex Gateway Service Pods

Restrict access to Flex Gateway runtime pods to ensure unauthorized personel cannot access Flex Gateway configuration data or tamper with the runtime.

> For Kubernetes deployments, use Kubernetes Role-based access control (RBAC) to restrict access to execute commands in the pod. For more information, see https://kubernetes.io/docs/reference/access-authn-authz/rbac/[Using RBAC Authorization].

Reviewer 2: ```suggestion
For Kubernetes deployments, restrict access to command execution in the pod by using Kubernetes Role-based access control (RBAC). For more information, see https://kubernetes.io/docs/reference/access-authn-authz/rbac/[Using RBAC Authorization].
```

In non-Kubernetes deployments, restrict access to the Flex Gateway virtual machine to open sessions.

[[secure-redis-shared-storage]]
== Secure Redis Shared Storage

> Flex Gateway caches both request data and runtime configurations (Connected Mode only) using Redis shared storage. An unauthorized user could obtain cached configuration data including Flex configuration or request processing caches or possibly disrupt the runtime. Secure the Redis service to ensure there is no unauthorized access.

Reviewer 2: ```suggestion
Flex Gateway uses Redis shared storage to cache request data and runtime configurations in Connected Mode only. An unauthorized user can obtain cached configuration data, such as Flex configuration or request processing caches, and can disrupt the runtime. You must secure the Redis service to prevent unauthorized access.
```

> To secure the Redis shared storage, see https://redis.io/docs/latest/operate/rs/security/recommended-security-practices/[Redis Recommended security practices]. Ensure you enable access control to limit Redis access from sources other than Flex Gateway. Deploy the Redis as close as possible to the Flex Gateway to further limit vulnerabilities.

Reviewer 2: ```suggestion
To secure the Redis shared storage, see https://redis.io/docs/latest/operate/rs/security/recommended-security-practices/[Redis Recommended security practices]. Enable access control to limit Redis access from sources other than Flex Gateway. Deploy the Redis as close as possible to the Flex Gateway deployment to further limit vulnerabilities.
```

[[protect-the-flow-of-incoming-traffic]]
== Protect the Flow of Incoming Traffic (API Consumer to Flex Gateway)

> Protecting the flow of incoming traffic from external attacks is dependent on the source of incoming traffic.

Reviewer 2: ```suggestion
The effectiveness of protecting incoming traffic from external attacks depends on the source of the traffic.```

> For requests coming from the public internet, ensure there are first line of defense systems to intercept the traffic before it reaches the Flex Gateway. While Flex Gateway does provides API level protection policies, it expects other general first line protections are in front of it. Ensure both:

Reviewer 2: ```suggestion
For requests coming from the public internet, configure a first line of defense system to intercept the traffic before it reaches the Flex Gateway. Flex Gateway provides API-level protection but relies on other general protections for the first line of defense. Ensure you have these protections:
```

> * Network level protections, such as Network ACLs, Network and DNS Firewalls, and other network protections available. Cloud providers or a third party may provide these protections.

Reviewer 2: ```suggestion
* Network level protections, like ACLs and DNS Firewalls. Cloud providers and third parties typically provide these protections.
```

* First line application level protections, such as threat protection against DDoS attacks and Web Application Firewall.

> NOTE: To find reference architecture detailing how to handle incoming requests from external sources, see xref:flex-architecture-uc2.adoc[] and xref:flex-architecture-uc4.adoc[]. 

Reviewer 2: ```suggestion
NOTE: For guidelines on handling incoming requests from external sources, see xref:flex-architecture-uc2.adoc[] and xref:flex-architecture-uc4.adoc[]. 
```

[[terminate-external-tls-at-point-of-entry]]
=== Terminate External TLS at Point of Entry

> Terminate external TLS connections at the firewall level to enable the firewall to perform any required L7 traffic security inspection. A valid Certificate Authority (CA) recognized by the consumers should provide the certificates at this point of entry.

Reviewer 2: ```suggestion
To enable the firewall to perform L7 traffic security inspections, terminate external TLS connections at the firewall level. A valid Certificate Authority (CA) that's recognized by the consumer typically provides the certificates at this point of entry.
```

> Use a separate set of certificates for internal communications between the firewall, Flex Gateway, and other services in the infrastructure. For this communication, use mTLS security with both certificates issued by a trusted authority (internal authority or PKI).

Reviewer 2: ```suggestion
Use a separate set of certificates for internal communication between the firewall, Flex Gateway, and other services in the infrastructure. For this communication, use mTLS security with both certificates issued by an internal authority or PKI.
```

> Certificates exposed to the internet are never used internally and certificates that are used internally are never exposed to the internet. Separating internal and external certificates prevents potential leakage of the internal credentials.

Reviewer 2: ```suggestion
Certificates exposed to the internet are never used internally, and certificates that are used internally are never exposed to the internet. Separating internal and external certificates prevents potential leakage of the internal credentials.
```

> You can use Flex Gateway as at both the runtime environment ingress layer or at the inner environment layer. For Flex Gateways at the ingress layer, limit the gateways scope to TLS termination, routing, and other basic protections. Push more CPU and memory intensive operations to inner Flex Gateways.

Reviewer 2: ```suggestion
You can use Flex Gateway at both the runtime environment ingress layer or at the inner environment layer. For Flex Gateways at the ingress layer, limit the gateways' scope to TLS termination, routing, and other basic protections. Push more CPU and memory intensive operations to inner Flex Gateways.
```

[[protect-apis-by-using-policies]]
=== Protect APIs by using Policies

> To further configure API protections, apply policies. Flex Gateway offers xref:policies-included-directory.adoc[Included Policies] and enables you to develop custom policies by using xref:pdk::policies-pdk-overview.adoc[].

Reviewer 2: ```suggestion
Apply policies to enhance API protections. Flex Gateway offers xref:policies-included-directory.adoc[Included Policies] and enables you to develop custom policies by using xref:pdk::policies-pdk-overview.adoc[].
```

[[govern-apis-by-using-api-governance]]
=== Govern APIs by using API Governance

Use xref:api-governance::index.adoc[API Governance] to ensure APIs are designed and deployed in line with MuleSoft, industry, and your organization's best practices. 

[[protect-the-upstream-service]]
== Protect the Upstream Service (Flex Gateway to API Provider)

Protecting the flow of traffic from Flex to the upstream service is dependent on the location of the upstream service.

[[protect-services-available-locally-to-flex-gateway]]
=== Protect Services Available Locally to Flex Gateway

> For services in the same network as Flex Gateway, security restrictions might be lower.

Reviewer 2: ```suggestion
Services that share the same network as Flex Gateway typically have lower security restrictions.
```

The organization must define whether plain or TLS connections are in place. If using TLS, configure mTLS via Flex Gateway Outbound TLS contexts. To learn more about configuring TLS, see xref:

* xref:flex-conn-tls-config.adoc[]
* xref:flex-local-tls-config.adoc[]

> Additionally, configure the upstream service to only accept traffic incoming from Flex Gateway. This protects the upstream service from other services in the same network. For Kubernetes, configure this using Network Policies. 

Reviewer 2: ```suggestion
To protect the upstream service from other services in the same network, configure the upstream service to accept traffic only from Flex Gateway. For Kubernetes, configure this using Network Policies. 
```

> NOTE: To find reference architecture detailing communication to an internal upstream service, see xref:flex-architecture-uc1.adoc[] and xref:flex-architecture-uc3.adoc[].

Reviewer 2: ```suggestion
NOTE: For guidelines on communicating to an internal upstream service, see xref:flex-architecture-uc1.adoc[] and xref:flex-architecture-uc3.adoc[].
```

[[protect-external-services]]
=== Protect External Services

> When protecting external services not in the same network, connect to the service using mTLS communication. Additionally, configure a trusted network link (private VPC link or other) between the networks to enable Flex Gateway to securely access the upstream service without exposing the service to the public internet. 

Reviewer 2: ```suggestion
When protecting external services not in the same network, connect to the service using mTLS communication. Configure a trusted network link, such as private VPC, between the networks to enable Flex Gateway to securely access the upstream service without exposing the service to the public internet. 
```

> NOTE: To find reference architecture detailing communication to an external upstream service, see xref:flex-architecture-uc2.adoc[] and xref:flex-architecture-uc4.adoc[].

Reviewer 2: ```suggestion
NOTE: For guidelines on communicating to an external upstream service, see xref:flex-architecture-uc2.adoc[] and xref:flex-architecture-uc4.adoc[].
```

> [[preform-regular-testing]]

Reviewer 2: ```suggestion
[[perform-regular-testing]]
```

> == Preform Regular Testing

Reviewer 2: ```suggestion
== Perform Regular Testing
```

> Perform regular penetration and vulnerability tests to identify new or regressed issues in your Flex Gateway deployment and exposed APIs. When testing, consider:

Reviewer 2: ```suggestion
Perform regular penetration and vulnerability tests to identify new or regressed issues in your Flex Gateway deployment and exposed APIs. Consider these when testing:
```

Reviewer 2: I think a word needs to be added between "these when" in the second sentence, but I wasn't sure what to suggest. Maybe "points" or "configurations" or "settings"?

* User access controls in Anypoint Platform that enable access to Flex registration in Runtime Manager, logs in Monitoring Center, and APIs in API Manager.
* Access to the deployed Flex Gateway that affects its function or data.
* Access to any configuration files used by the Flex runtime, especially the registration file.
* Access to the Redis shared storage service.
* Access to APIs deployed on the Flex Gateway. 

> * Access to the upstream services protecting by the Flex Qateway APIs.

Reviewer 2: ```suggestion
* Access to the upstream services protected by the Flex Gateway APIs.
```

For Kubernetes deployments, use Kubernetes Role-based access control (RBAC) to restrict access to execute commands in the pod. For more information, see https://kubernetes.io/docs/reference/access-authn-authz/rbac/[Using RBAC Authorization].

In non-Kubernetes deployments, restrict access to the Flex Gateway virtual machine to open sessions.

[[secure-redis-shared-storage]]
== Secure Redis Shared Storage

Flex Gateway caches both request data and runtime configurations (Connected Mode only) using Redis shared storage. An unauthorized user could obtain cached configuration data including Flex configuration or request processing caches or possibly disrupt the runtime. Secure the Redis service to ensure there is no unauthorized access.

To secure the Redis shared storage, see https://redis.io/docs/latest/operate/rs/security/recommended-security-practices/[Redis Recommended security practices]. Ensure you enable access control to limit Redis access from sources other than Flex Gateway. Deploy the Redis as close as possible to the Flex Gateway to further limit vulnerabilities.

[[protect-flex-from-incoming-traffic]]

> == Protect Flex From Incoming Traffic (API Consumer to Flex Gateway)

Reviewer: We are protecting the incoming traffic flow. We are not protecting Flex from the incoming traffic. 
Flex expects incoming traffic by definition as it is its main function to do so.

> The protections necessary to protect Flex Gateway from incoming traffic is dependent on the source of incoming traffic.

Reviewer: I would not say this. We are not trying to protect Flex from the incoming traffic. Rather we are protecting the incoming traffic flow from external attacks.

Certificates exposed to the internet are never used internally and certificates that are used internally are never exposed to the internet. Separating internal and external certificates prevents potential leakage of the internal credentials.

You can use Flex Gateway as at both the runtime environment ingress layer or at the inner environment layer. For Flex Gateways at the ingress layer, limit the gateways scope to TLS termination, routing, and other basic protections. Push more CPU and memory intensive operations to inner Flex Gateways.

[[protect-apis-by-using-policies]]
=== Protect APIs by using Policies

To further configure API protections, apply policies. Flex Gateway offers xref:policies-included-directory.adoc[Included Policies] and enables you to develop custom policies by using xref:pdk::policies-pdk-overview.adoc[].

[[govern-apis-by-using-api-governance]]
=== Govern APIs by using API Governance

Use xref:api-governance::index.adoc[API Governance] to ensure APIs are designed and deployed in line with MuleSoft, industry, and your organization's best practices. 

[[protect-the-upstream-service]]
== Protect the Upstream Service (Flex Gateway to API Provider)

> Protecting the upstream service is 

Reviewer: Missing text here?

> The protections necessary to protect Flex Gateway from incoming traffic is dependent on the source of incoming traffic.

Reviewer: I don't think this makes sense here.  We are protecting traffic between flex and the upstream, not the incoming traffic.
